### ISSUE TITLE
Handling repeated subtitles from YouTube?

### ISSUE THREAD

**Issue by edrozenberg** [iina/iina#4120]
I downloaded some auto-subtitles from YouTube recently, and the previous subtitle is shown with each new subtitle in IINA, for ex:
```
Hello
I'm Here

I'm here
how are you?

how are you?
I'm doing well
```

Snippet of the downloaded subtitle `vtt`:

```
WEBVTT
Kind: captions
Language: en

00:00:00.000 --> 00:00:01.550 align:start position:0%
 
dear <00:00:00.340><c>friends </c><00:00:00.680><c>Dear </c><00:00:01.020><c>viewers,</c>

00:00:01.550 --> 00:00:01.560 align:start position:0%
dear friends Dear viewers,
 

00:00:01.560 --> 00:00:03.590 align:start position:0%
dear friends Dear viewers,
listeners <00:00:01.850><c>and, </c><00:00:02.140><c>of </c><00:00:02.430><c>course, </c><00:00:02.720><c>subscribers </c><00:00:03.010><c>of </c><00:00:03.300><c>the</c>

00:00:03.590 --> 00:00:03.600 align:start position:0%
listeners and, of course, subscribers of the
```

YouTube correctly displays only one subtitle at a time. Not sure that IINA can deal with this issue, since the duplicated subs are in the `vtt` downloaded from YouTube.

**What you want IINA to do:**
Display only one subtitle line, without the previous subtitle.

**What IINA does currently:**
Displays the current subtitle and the previous subtitle at the same time.

**Why you think this should be added:**
Showing previous + current subtitle is confusing and hard to follow.

**Examples of other projects that have something similar:**

**Comment by low-batt** [comment]
The IINA issue template contains this checkbox:
> [ ] MPV does not have this problem.

The reason for asking this question is that for video/audio playback IINA uses `libmpv` from the [mpv](https://mpv.io/) project. If the problem reproduces using the mpv player then for IINA it is an "upstream" problem. The issue must be reported to the mpv project at: [mpv issues](https://github.com/mpv-player/mpv/issues). Once that project releases a new version with a fix, IINA must be updated to use the latest version of the library from mpv.

In this case it is mpv that is rendering subtitles into the video. My expectation is that the mpv player will exhibit the same behavior as IINA. Would need a URL to a YouTube video that reproduces the issue to test that.

As for what is the proper way to render the subtitles, that is specified in detail in the document [WebVTT: The Web Video Text Tracks Format](https://www.w3.org/TR/webvtt1/#normative). This type of subtitle file must adhere to that specification. As you can see from that document this form of subtitle file has a lot of capabilities.

**Comment by edrozenberg** [comment]
Thank you. I didn't see the mpv template item for some reason. Same issue with `mpv 0.35.0` (mpv.io) installed via homebrew cask. The subtitle file from YouTube contains the duplicate subtitles in each subtitle entry. Closing for now.

**Comment by low-batt** [comment]
Thanks for confirming mpv also exhibits the same problem. When you report it to the mpv project post the issue number here so the IINA project can watch for a fix.

If you don't have time to report the problem to mpv you can post a link to a YouTube video that exhibits the problem here and I will try and reproduce the problem and enter an issue with the mpv project.

The IINA convention for such issues is to mark the issue with the "upstream" label and leave the issue open until the problem has been fixed in the IINA dependency and a new version of IINA has been released with the fix. This way it is more likely that other users looking to report the same will find the open issue and not enter a duplicate.

**Comment by solidogen** [comment]
Why was this closed? This is a major bug.

**Comment by low-batt** [comment]
What are GitHub [issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/learning-about-issues/about-issues)?
- A list of problems in the current release of an application?
- A list of tasks for developers to do?

The linked GitHub documentation does not spell it out, but GitHub behavior shows GitHub views issues as tasks for developers to do. When changes in a pull request associated with an issue are merged, a GitHub bot closes the issue, even though the latest release of the application is still broken and the fix is only in the version being developed.

IINA tries to use issues like a bug database representing the state of the current release. But the GitHub bots are winning the war, so we are no longer re-opening issues the bots close.

With respects to this specific issue…

Like a lot of open source apps, IINA is built on top of libraries from other projects. For audio/video and subtitle rendering IINA embeds a library from the [mpv project](https://github.com/mpv-player/mpv).  It is `libmpv` that (from IINA code's standpoint) reads subtitle files, processes them and embeds the subtitles in the video frames.

This is what IINA refers to as an "upstream" problem. The root cause of the defect is in code from other projects that IINA uses. The IINA project cannot fix the problem. The problem must be reported to the project associated with the broken code. That project must fix the problem and generate a new release. Then the IINA project must upgrade to the version of the library with the fix.

I expect @edrozenberg closed this issue once it had been determined that the root cause is not in IINA code, as that is a common GitHub convention. I re-opened it so that users are more likely to find this issue and not enter a duplicate.

When it comes to where the root cause of this problem is, I can't tell you where the problem lies. From a quick search of [mpv issues](https://github.com/mpv-player/mpv/issues), it seems mpv uses [FFmpeg](https://ffmpeg.org/) for processing subtitles in [WebVTT](https://en.wikipedia.org/wiki/WebVTT) format and there are problems in FFmpeg with its WebVTT support. There was also mention of [yt-dlp](https://github.com/yt-dlp/yt-dlp) not producing correct WebVTT files when downloading from YouTube.

**Comment by solidogen** [comment]
Sorry - I acted without properly inspecting the issue first.

I inspected the output of my subtitle files downloaded by `yt-dlp` and indeed the issue is either on `YouTube` or `yt-dlp` side. I assume author of this issue also used `yt-dlp` to download his videos with subtitles.

Indeed `IINA` and `mpv` work fine in this manner and there is nothing wrong with closing this issue when I look at it now.

Actual issue and workarounds can be seen here:
https://github.com/yt-dlp/yt-dlp/issues/1734

**Comment by low-batt** [comment]
@solidogen Thanks so much for identifying the `yt-dlp` issue.

I'm going to leave this issue open for now. The `yt-dlp` issue mentions using `yt-dlp` plugins as a solution. If that ends up being the fix, IINA might need some changes to allow use of `yt-dlp` plugins.
